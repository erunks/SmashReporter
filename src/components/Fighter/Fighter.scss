@import '../../styles/base.scss';

.fighter {
  display: flex;
  position: relative;
}

.fighterCard {  
  background-repeat: no-repeat;
  background-size: cover;
  border-right: 2px solid;
  height: 55px;
  max-width: 100px;
  position: relative;
  width: 90px;

  &:hover {
    z-index: 2;

    .selection {
      @include visible(inline);
    }
  }

  .selection {
    &.player-one {
      @include blink-animation(primary-blink);
      border-color: $primary-selection-color;

      .bottomRight,
      .topLeft {
        @include blink-animation(primary-corner-blink);
      }

      .bottomRight {
        border-top-color: $primary-selection-color;
      }

      .topLeft {
        border-bottom-color: $primary-selection-color;
      }
    }
    &.player-two {
      @include blink-animation(secondary-blink);
      border-color: $secondary-selection-color;

      .bottomRight,
      .topLeft {
        @include blink-animation(secondary-corner-blink);
      }

      .bottomRight {
        border-top-color: $secondary-selection-color;
      }

      .topLeft {
        border-bottom-color: $secondary-selection-color;
      }
    }

    @include absolute-position;
    @include hidden;

    border: 4px solid;
    height: 100%;
    width: 100%;

    .bottomRight,
    .topLeft {
      height: 0;
      position: absolute;
      width: 0;
    }

    .bottomRight {
      border-right: 12px solid transparent;
      border-top: 12px solid;
      left: 0;
      top: 0;
    }

    .topLeft {
      border-bottom: 12px solid;
      border-left: 12px solid transparent;
      bottom: 0;
      right: 0;
    }
  }

  .namePlate {
    background-color: black;
    height: 30px;
    max-width: 100%;
  }

  .name {
    color: white;
    font-size: 12px;
    padding: 0.25em 2px;
    margin: 0;
    overflow: hidden;
    text-align: center;
    text-overflow: ellipsis;
    white-space: nowrap;
    width: 100%;
  }
}

@keyframes primary-blink {
	0% {
    border-color: $primary-selection-color;
  }
  
  50% {
    border-color: white;
  }

  100% {
    border-color: $primary-selection-color;
  }
}

@keyframes primary-corner-blink {
	0% {
    border-bottom-color: $primary-selection-color;
    border-top-color: $primary-selection-color;
  }
  
  50% {
    border-bottom-color: white;
    border-top-color: white;
  }

  100% {
    border-bottom-color: $primary-selection-color;
    border-top-color: $primary-selection-color;
  }
}

@keyframes secondary-blink {
	0% {
    border-color: $secondary-selection-color;
  }
  
  50% {
    border-color: white;
  }

  100% {
    border-color: $secondary-selection-color;
  }
}

@keyframes secondary-corner-blink {
	0% {
    border-bottom-color: $secondary-selection-color;
    border-top-color: $secondary-selection-color;
  }
  
  50% {
    border-bottom-color: white;
    border-top-color: white;
  }

  100% {
    border-bottom-color: $secondary-selection-color;
    border-top-color: $secondary-selection-color;
  }
}
