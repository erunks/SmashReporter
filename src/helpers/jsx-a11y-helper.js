export function buttonize(action) {
  return {
    role: 'button',
    onClick: action,
    onKeyDown: (event) => {
      // Handles Enter/Return pressed
      if (event.keycode === 13) {
        action(event);
      }
    }
  };
}

export default { buttonize };
