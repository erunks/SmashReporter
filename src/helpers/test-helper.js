import React from 'react';
import { render } from '@testing-library/react';
import { mount, shallow } from 'enzyme';
import { IKContext } from 'imagekitio-react';
import { RecoilRoot } from 'recoil';

const wrapper = (component, options = {
  initializeState: () => {},
}) => {
  const initialStateMethod = options.initializeState;

  return (
    <IKContext
      publicKey="public_DfZPshOEvWdxzvxS8HGNWPJCo70="
      urlEndpoint="https://ik.imagekit.io/ozcerk4wii"
    >
      <RecoilRoot initializeState={ initialStateMethod }>
        { component }
      </RecoilRoot>
    </IKContext>
  );
};

const mountWrapper = (component, options) => mount(wrapper(component, options));
const shallowWrapper = (component, options) => shallow(wrapper(component, options));
const wrappedRender = (component, options = {}) => {
  const initialStateMethod = options && options.initializeState;

  const Wrapper = ({ children }) => (
    <IKContext
      publicKey="public_DfZPshOEvWdxzvxS8HGNWPJCo70="
      urlEndpoint="https://ik.imagekit.io/ozcerk4wii"
    >
      <RecoilRoot initializeState={ initialStateMethod }>
        { children }
      </RecoilRoot>
    </IKContext>
  );

  return render(component, { wrapper: Wrapper, ...options });
};


export * from '@testing-library/react';

export {
  mountWrapper,
  shallowWrapper,
  wrappedRender
};
