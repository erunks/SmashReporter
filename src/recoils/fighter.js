import { atom, selector } from 'recoil';
import findIndex from 'lodash/findIndex';
import slice from 'lodash/slice';

export const fighterState = atom({
  key: 'fighterState',
  default: [{}]
});

export const orderedFighters = selector({
  key: 'getFighters',
  get: ({ get }) => {
    const fighters = get(fighterState);
    const start = findIndex(fighters, { name: 'Mii Fighter Brawler' });
    if (start) {
      const before = slice(fighters, 0, start);
      const pulled = slice(fighters, start, start + 3);
      const after = slice(fighters, start + 3);
      return [...before, ...after, ...pulled];
    } else {
      return fighters;
    }
  },
});

export const selectedFightersState = atom({
  key: 'selectedFightersState',
  default: ['random','random']
});

export default {
  fighterState,
  orderedFighters,
  selectedFightersState
};
