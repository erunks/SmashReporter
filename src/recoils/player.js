import { atom } from 'recoil';

export const selectingPlayerState = atom({
  key: 'selectingPlayerState',
  default: 0
});

export default {
  selectingPlayerState
};
