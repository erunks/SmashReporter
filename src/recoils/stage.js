import { atom, selector } from 'recoil';

export const selectedStageState = atom({
  key: 'selectedStageState',
  default: 1
});

export const getSelectedStage = selector({
  key: 'getSelectedStage',
  get: ({ get }) => (get(selectedStageState) - 1),
});

export default {
  getSelectedStage,
  selectedStageState
};
