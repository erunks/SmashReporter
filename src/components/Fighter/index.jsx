import React from 'react';
import propTypes from 'prop-types';
import { URLS } from '../../lib/imageAssets';

import './Fighter.scss';

const Fighter = (props) => {
  const fighterAssetName = (name) => {
    let assetName = name.replace(/[- ]/g, '_');
    assetName = assetName.replace(/&/g, 'and');
    assetName = assetName.replace(/\./g, '');
    return assetName.toLowerCase();
  };

  return (
    <div className="fighter">
      <div className="fighterCard">
        <img src={URLS.fighterThumbnailIcon(fighterAssetName(props.name))} alt={`fighter: ${props.name}`} />
        <div className="selection">
          <div className="topLeft"></div>
          <div className="bottomRight"></div>
        </div>
      </div>
    </div>
  );
};

const {
  bool,
  number,
  oneOfType,
  string
} = propTypes;

Fighter.propTypes = {
  dlc: oneOfType([
    bool,
    number
  ]).isRequired,
  id: number.isRequired,
  name: string.isRequired
};

export default Fighter;
