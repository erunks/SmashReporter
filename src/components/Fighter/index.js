import React from 'react';
import propTypes from 'prop-types';
import { useRecoilState, useRecoilValue } from 'recoil';
import { selectedFightersState } from 'recoils/fighter';
import { selectingPlayerState } from 'recoils/player';
import { buttonize } from 'helpers/jsx-a11y-helper'
import { IKImage } from 'imagekitio-react';
import { URLS } from 'lib/imageAssets';

import './Fighter.scss';

const Fighter = (props) => {
  const [selectedFighters, setSelectedFighters] = useRecoilState(selectedFightersState);
  const selectingPlayer = useRecoilValue(selectingPlayerState);

  const fighterAssetName = (name) => {
    let assetName = name.replace(/[- ]/g, '_');
    assetName = assetName.replace(/&/g, 'and');
    assetName = assetName.replace(/\./g, '');
    return assetName.toLowerCase();
  };

  const setFighter = () => {
    const newSelectedFighters = [...selectedFighters];
    newSelectedFighters[selectingPlayer] = asset_name;
    setSelectedFighters(newSelectedFighters);
  };

  const asset_name = props.asset_name || fighterAssetName(props.name);
  const fighter_thumbnail_icon = props.asset_url || URLS.fighterThumbnailIcon(asset_name);

  return (
    <div className="fighter" { ...buttonize(() => setFighter()) }>
      <div className="fighterCard" alt={`fighter: ${props.name}`}>
        <IKImage
          src={fighter_thumbnail_icon}
          alt={`fighter: ${props.name}`}
          loading="lazy"
          lqip={{ active: true }}
        />
        <div className={`selection ${selectingPlayer === 0 ? 'player-one' : 'player-two'}`}>
          <div className="topLeft"></div>
          <div className="bottomRight"></div>
        </div>
      </div>
    </div>
  );
};

const {
  bool,
  number,
  oneOfType,
  string
} = propTypes;

Fighter.propTypes = {
  dlc: oneOfType([
    bool,
    number
  ]).isRequired,
  id: number.isRequired,
  name: string.isRequired
};

export default Fighter;
