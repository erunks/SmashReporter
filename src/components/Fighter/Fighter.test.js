import React from 'react';
import { mountWrapper, wrappedRender } from 'helpers/test-helper';
import { selectingPlayerState } from 'recoils/player';
import Fighter from '.';

describe('<Fighter>', () => {
  const props = {
    dlc: 0,
    id: 1,
    name: 'Mario'
  };

  it('should match the snapshot', () => {
    const container = mountWrapper(
      <Fighter {...props} />,
      { initializeState: (snap) => { snap.set(selectingPlayerState, 0); } }
    );
    expect(container.html()).toMatchSnapshot();
  });

  test('when the selectingPlayer is player two, the selection should have the accompanying class', () => {
    wrappedRender(
      <Fighter {...props} />,
      { initializeState: (snap) => { snap.set(selectingPlayerState, 1); } }
    );

    const selection = document.querySelector('.player-two');
    expect(selection).toBeInTheDocument();
  });
});
