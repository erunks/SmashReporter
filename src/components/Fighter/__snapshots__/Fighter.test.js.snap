// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Fighter> should match the snapshot 1`] = `"<div class=\\"fighter\\" role=\\"button\\"><div class=\\"fighterCard\\" alt=\\"fighter: Mario\\"><img alt=\\"fighter: Mario\\" src=\\"https://ik.imagekit.io/ozcerk4wii/fighter/thumb_a/mario.png?ik-sdk-version=react-1.0.8&amp;tr=q-20%2Cbl-6\\"><div class=\\"selection player-one\\"><div class=\\"topLeft\\"></div><div class=\\"bottomRight\\"></div></div></div></div>"`;
